src/clj/puppetlabs/services/master/file_serving.clj

```diff
@@ -0,0 +1,364 @@
+(ns puppetlabs.services.master.file-serving
+  (:require [bidi.bidi :as bidi]
+            [clj-yaml.core :as yaml]
+            [clojure.java.io :as io]
+            [clojure.string :as str]
+            [digest :as digest]
+            [me.raynes.fs :as fs]
+            [puppetlabs.i18n.core :as i18n]
+            [puppetlabs.ring-middleware.utils :as middleware-utils]
+            [ring.util.response :as rr])
+  (:import com.sun.security.auth.module.UnixSystem
+           [java.nio.file Files FileSystems FileVisitOption FileVisitResult LinkOption Paths SimpleFileVisitor]))
+
+(def follow-links
+  (into-array LinkOption []))
+
+(defn as-path
+  [string & more]
+  (Paths/get string (into-array String more)))
+
+(defn path-exists?
+  [path]
+  (Files/exists path (into-array LinkOption [])))
+
+(as-path "/home/patrick" "data")
+
+(defn get-checksum
+  [file algorithm]
+  (case algorithm
+    "md5" (str "{md5}" (digest/md5 (io/file file)))
+    "sha256" (str "{sha256}" (digest/sha-256 (io/file file)))
+    :default (throw (Exception. "Unsupported digest"))))
+
+(defn read-attributes
+  [path checksum-type ignore-source-permissions]
```

Comment:
`source_permissions` is an option on file resources that indicates whether the client wants to receive the actual on disk permissions of the files or whether it gets a default set of permissions and uid/gid of the server process. It is deprecated, but still used for pluginsync (where it's generally set to `ignore` which gets converted to `true` in this representation). I have no idea why anyone ever wanted to use it.

Reply:
so we default to ignoring..? which uses the default set of permissions? That seems really weird to me.. Not saying you did something wrong, just trying to understand.

Reply:
People surprisingly like being able to copy the file mode when managing a recursive directory resource (https://tickets.puppetlabs.com/browse/PUP-10253) Copying the owner and group is nonsensical but currently supported, and hopefully will be removed at some point. Puppet agents use `source_permissions => use` for the pluginfacts mount, because we need to preserve the executable-ness of external executable facts.